request/executor.go

```diff
@@ -7,44 +7,85 @@ import (
 )
 
 // ExecuteRequest executes an HTTP request based on the specified options.
-func ExecuteRequest(request Request, profileNames []string, variables map[string]string) (*ExecutedRequestResponse, error) {
-	httpRequest, configuredRequest, httpRequestErr := BuildRequest(request, profileNames, variables)
-	if httpRequestErr != nil {
-		return nil, httpRequestErr
-	}
+func ExecuteRequest(request Request, profileNames []string, variables map[string]string) ([]ExecutedRequestResponse, error) {
+	toExecute := make([]Request, 1)
+	toExecute[0] = request
 
-	client := &http.Client{}
-	httpResponse, httpResponseErr := client.Do(httpRequest)
-	if httpResponseErr != nil {
-		return nil, httpResponseErr
+	client := &http.Client{
+		// Do not auto-follow redirects
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
 	}
 
-	bodyBytes, readErr := ioutil.ReadAll(httpResponse.Body)
+	result := make([]ExecutedRequestResponse, 0)
 
-	if readErr != nil {
-		return nil, readErr
-	}
+	for {
+		executing := toExecute[0]
+		toExecute = toExecute[1:]
 
-	body := ""
-	if len(bodyBytes) != 0 {
-		body = string(bodyBytes)
-	}
+		httpRequest, configuredRequest, httpRequestErr := BuildRequest(executing, profileNames, variables)
+		if httpRequestErr != nil {
+			return nil, httpRequestErr
+		}
 
-	headers := make(map[string][]string)
-	for k, vs := range httpResponse.Header {
-		headers[k] = append(headers[k], vs...)
-	}
+		httpResponse, httpResponseErr := client.Do(httpRequest)
+		if httpResponseErr != nil {
+			return nil, httpResponseErr
+		}
+
+		bodyBytes, readErr := ioutil.ReadAll(httpResponse.Body)
+
+		if readErr != nil {
+			return nil, readErr
+		}
+
+		body := ""
+		if len(bodyBytes) != 0 {
+			body = string(bodyBytes)
+		}
 
-	response := Response{
-		StatusCode: httpResponse.StatusCode,
-		Status:     httpResponse.Status,
-		Headers:    headers,
-		Body:       body,
-		Protocol:   fmt.Sprintf("%d.%d", httpResponse.ProtoMajor, httpResponse.ProtoMinor),
+		headers := make(map[string][]string)
+		for k, vs := range httpResponse.Header {
+			headers[k] = append(headers[k], vs...)
+		}
+
+		response := Response{
+			StatusCode: httpResponse.StatusCode,
+			Status:     httpResponse.Status,
+			Headers:    headers,
+			Body:       body,
+			Protocol:   fmt.Sprintf("%d.%d", httpResponse.ProtoMajor, httpResponse.ProtoMinor),
+		}
+
+		result = append(result, ExecutedRequestResponse{
+			Request:  *configuredRequest,
+			Response: response,
+		})
+
+		if response.StatusCode == http.StatusMovedPermanently || response.StatusCode == http.StatusFound || response.StatusCode == http.StatusSeeOther {
```

Comment:
[From the `http.Client#Do` documentation](https://golang.org/pkg/net/http/#Client.Do)
> If the server replies with a redirect, the Client first uses the CheckRedirect function to determine whether the redirect should be followed. If permitted, a 301, 302, or 303 redirect causes subsequent requests to use HTTP method GET (or HEAD if the original request was HEAD), with no body. A 307 or 308 redirect preserves the original HTTP method and body, provided that the Request.GetBody function is defined. The NewRequest function automatically sets GetBody for common standard library body types.

Are you planning on supporting `307` (`http.StatusTemporaryRedirect`) and `308` (`http. StatusPermanentRedirect`)?

Reply:
I thought about it. But then reading all of these different types of redirect I figured that they need more attention. Some of them require you to send the same request just to a different URL (same method, body, etc), some of them mean just "go somewhere else, whatever you want is not here".

So right now, I'm choosing the simple path and implementing the most widely used ones (I've never seen a 307 nor 308, only 301s and 302s) and I'm also not choosing to implement them 100% as they should be.

The main goal of the issue was just to break the redirects apart internally so that there's a way to tap into the data between the requests.